Develop

niesi opened the pull request:
Hallo Stefan!

Das hier ist jetzt die versprochene Mail, oder wie auch immer man es nennen mag! Wie wir in der Umit besprochen haben sende ich dir ein paar Überlegungen und Änderungen, die ich in Code gemacht habe.
Da ich mit GitHub und ähnlichem keine Erfahrung habe, verzeihe mir bitte wenn mein pull request wahrscheinlich nicht  so ist, wie er sein sollte.... :)
Im Anhang habe ich das komplette Projekt gezippt, mit zwei Beispielen vom Einfach- und vom Dreifachpendel die nicht der matplotlib GUI von Marcus funktionieren. Lieder funktionieren noch nicht alle regimes.

Also einige Ideen mit Code:

DeadTimeSensor:
Es wird das komplette "input signal" ausgegeben, nur die gewählten Zustände werden verzögert.

class DeadTimeSensor(Sensor):
    """
    Sensor that adds a measurement delay
    """

```
public_settings = OrderedDict([("states to delay", [0,1,2]),
                               ("delay", 1)])

def __init__(self, settings):
    settings.update([("input signal", "system_state")])
    Sensor.__init__(self, settings)
    self._storage = None

def _measure(self, value):
    if self._storage is None:
        self._storage = []
        for i in range(int(self._settings["delay"])):
            # todo simon append first input, not zeros!!!
            self._storage.append(value)
            # if self._settings["delay"]
            # self._storage =  * self._settings["delay"]

    # add new measurement
    measurement = value.copy()
    self._storage.append(value)

    # get delayed measurements
    meas_delayed = self._storage.pop(0)

    # replace selected values for delayes values

    for i in self._settings["states to delay"]:
        measurement[i] = meas_delayed[i]

    return measurement
```

In simulation_core
Rückgabewerte der Module können auch dics sein. Es wird das komplette dictionary gespeichert.
 und anschießend in den Ergebnissen angezeigt. z.b. wenn von der Regelung die Verstärkung und die Stellgröße interresant ist. 

```
def _store_values(self):
    """
    store all values of finished integration step
    """
    # todo Simon
    for key, val in self._current_outputs.items():
        if key in self._storage:
            if isinstance(val,dict):
                for subkey in val.keys():
                    self._storage[key][subkey].append(val[subkey])
            else:
                self._storage[key].append(val)
        else:
            if isinstance(val, dict):
                self._storage[key] = {}
                for subkey in val.keys():
                    self._storage[key][subkey] = [val[subkey]]
            else:
                self._storage[key]= [val]
    return
```

In Simulation Interface "_get_settings"
Es kann alles direkt in den settings eingegeben werden, das numpy-kompatibel oder ein dict ist (z.B.) Stellgrößenbeschränkungen für Pytrajektory

```
def _get_settings(self, model, module_name):
    item = model.findItems(module_name).pop(0)

    settings = OrderedDict()
    for row in range(item.rowCount()):
        property_name = str(item.child(row, 0).text())
        val_str = str(item.child(row, 1).text())

        #TODO Simon allows every numpy compatible input
        if self.is_number(val_str):
            prop_val = float(val_str)
        elif '{' in val_str:
            exec 'prop_val =' + val_str
        elif '[' in val_str or 'pi' in val_str or 'np.' in val_str:
            exec 'prop_val = np.array(' + val_str + ')'
        elif val_str == 'None' or val_str == 'none':
            prop_val = None
        else:
            # string
            prop_val = val_str

        settings.update({property_name: prop_val})

    return settings

# TODO Simon
def is_number(self, s):
    try:
        complex(s)  # for int, long, float and complex
    except ValueError:
        return False

    return True
```

In "_postprocessing"
ausgabe der als dict gespeicherten results. Die Solltrajectorie stellt bei mir immer den gesamten Zustand dar, nicht wie original ein Zustand und deren Ableitungen. Deshalb functioniert das mit dem "control_error"

 def _postprocessing(self):
        """
        calculation of some basic metrics for quick judging of simulation results
        """

```
    # TODO make this able to calc error for vector-like model output
    # todo Simon Done!
    # control and observer error
    if "Trajectory" in self._sim_data["results"]:
        c_error = self._get_result_by_name("Trajectory").squeeze() - self._get_result_by_name("Model").squeeze()
        self._sim_data['results'].update(control_error=c_error)
    if "Observer" in self._sim_data["results"]:
        o_error = self._get_result_by_name("Solver").squeeze() - self._get_result_by_name("Observer").squeeze()
        self._sim_data['results'].update(observer_error=o_error)

    #if a dict is returend, add modules if needed
    for result in self._sim_data["results"].keys():
        if result == 'Controller' and isinstance(self._sim_data["results"][result],dict):
            for key in self._sim_data["results"][result].keys():
                self._sim_data['results'][result + ': ' + key] = np.array(self._sim_data["results"][result][key])
            del self._sim_data["results"][result]

    self.save_file()
```

"save_file" speichert eine Datei ab, genau so wie ich es am Versuchsstand dann brauche. vielleicht ist es irgendwie nützlich:

 def save_file(self):
        # save trajektory
        controller_path = self.create_dir('Trajectory')
        time_stamp = time.strftime('%d%m%y_%H%M%S')  # DDMMYY_hhmmss
        file_name = controller_path + os.path.sep + 'trajectory_' + time_stamp + ".csv"
        header = 'File generated with PyMoskito and Pytrajektory\n' \
                 'Simon Niederkofler\n'

```
    if self._sim_data['modules']['Trajectory']['type'] == 'PyTrajectory':
        to_log = ['Trajectory', 'Feedforward', 'Controller', 'Solver']

        time_a = self._sim_data['modules']['Trajectory']['time a']
        time_b = self._sim_data['modules']['Trajectory']['time b']
        time_array = self.to_column(np.linspace(time_a, time_b, (time_b - time_a) * self._sim_data['modules']['Solver']['measure rate'] + 1))
        samples = time_array.shape[0]

        time_array_sim = self._sim_data["results"]['time']

        x_array = self.to_column(self._sim_data["results"]['Trajectory'])
        x_array = self.interp_array(time_array, time_array_sim, x_array)
        x_array = self.reshape_4to8(x_array)

    elif self._sim_data['modules']['Trajectory']['type'] == 'Setpoint':
        to_log = ['Trajectory', 'Controller', 'Solver']

        time_array = np.array([[0]]) #only one point 2D!!
        x_array = self.reshape_4to8(self._sim_data['modules']['Trajectory']['Setpoint'])
        time_array_sim = np.array([0])
        samples = 1

    for modul in to_log:
        if modul in self._sim_data['modules'].keys():
            header += '\n' + modul + ':\n'
            for item, value in self._sim_data['modules'][modul].items():
                header += '    ' + item + '=' + str(value) + '\n'

    if 'Feedforward' in self._sim_data["results"].keys() and 'Feedforward' in to_log:
        if samples == 1:
            u_array = self.to_column(self._sim_data["results"]['Feedforward'])[0,:].reshape(-1,1)
        else:
            u_array = self.to_column(self._sim_data["results"]['Feedforward'])
            u_array = self.interp_array(time_array,time_array_sim,u_array)
    else:
        u_array = np.zeros([samples, 1])

    if 'Controller: k' in self._sim_data["results"].keys() and 'Controller' in to_log:
        if samples == 1:
            k_array = self.to_column(self._sim_data["results"]['Controller: k'])[0,:].reshape(-1,1)
        else:
            k_array = self.to_column(self._sim_data["results"]['Controller: k'])
            k_array = self.interp_array(time_array, time_array_sim, k_array)
        k_array = self.reshape_4to8(k_array)
    else:
        k_array = np.zeros([samples,0])

    header += '\n\nTime[s];'
    for i in range(u_array.shape[1]):
        header += 'u' + str(i+1) + ';'
    for i in range(x_array.shape[1]):
        header += 'x' + str(i+1) + ';'
    for i in range(k_array.shape[1]):
        header += 'k' + str(i+1) + ';'

    np.savetxt(file_name, np.hstack((time_array,u_array,x_array,k_array)), fmt='%.5f', delimiter=';',header=header)



def create_dir(self, dir_name):
    path = os.getcwd() + os.path.sep + dir_name
    if not os.path.exists(path) or not os.path.isdir(path):
        os.mkdir(path)
    return path

def to_column(self, vector):
    # type: (np.array) -> np.array
    dim = len(vector.shape)
    if dim == 1:
        #row vector
        return np.array([vector]).T
    elif dim == 2:
        if vector.shape[1] > vector.shape[0]:
            return vector.T
        else:
            return vector
    else:
        self._logger.error("Dimension > 2!")

def interp_array(self,x, xp, yp_array):
    # type: (np.array, 1D np.array, np.array) -> np.array
    y = np.zeros([len(x), yp_array.shape[1]])
    for i in range(yp_array.shape[1]):
        y[:,[i]] = np.interp(x,xp.squeeze(),yp_array[:,i].squeeze())
    return y

def reshape_4to8(self, array):
    if len(array.shape) == 1:
        col = array.shape[0]            #[0,0,0,0]
    elif len(array.shape) == 2:         #[[[0],[0],[0],[0]]]
        col = array.shape[1]
    else:
        raise ValueError('Dimension > 2')
    if col == 4:
        array_8 = np.zeros([array.shape[0], 8])
        array_8[:, [0]] = array[:, [0]]
        array_8[:, [3]] = array[:, [1]]
        array_8[:, [4]] = array[:, [2]]
        array_8[:, [7]] = array[:, [3]]
        return array_8
    else:
        return array.reshape(-1,8)
```

In "simulaiotn_modules" in "calc_output" wenn keine Trajektorie vorgegeben ist, wird stattdessen null verwendet.

```
def calc_output(self, input_vector):
    # TODO Simon: Added Possibility for "None"-Input
    if self._settings["input_type"] == "none" or self._settings["input_type"] == "None":
        input_values = [0]
    else:
        input_values = next((input_vector[src] for src in self.input_sources
                             if src == self._settings["input_type"]), None)
        input_values = input_values.squeeze()
    if input_values is None:
        raise ControllerException("Selected Input not available")

    #todo Simon: option for no trajectory
    desired_values = {}
    if "Trajectory" in input_vector:
        desired_values['x'] = input_vector["Trajectory"]
    else:
        desired_values['x'] = np.zeros(self._model._settings['state_count'])

    if "Feedforward" in input_vector:
        desired_values['u'] = input_vector["Feedforward"]
    else:
        desired_values['u'] = np.zeros(self._model._settings['input_count'])


    return self._control(input_values, desired_values, input_vector["time"])
```

Ich habe generell bei allen Modulen "input" als Einstellung bebehalten. So kann man ganz einfach das modulare Konzept beibehalten, z.B. wenn nach "Sensor", "Disturbance" oder "system_state" geregelt werden soll.....
Irgendwie funktioniert auch die Implementierung von Pytrajektory, sollte aber doch sauber mit eigenem Thread usw. gemacht werden. Für meine Zwecke reicht es so. .D
Habe in den Modellen verschiedene Parameter für die Auslegung der Trajektorie und die Anschließende Simulation vorgesehen. Gerade beim Dreifachpendel kann man so interessante Effekte simulieren. so kann kein kleinen Parameterabweichungen die untere Ruhelage instabil werden.

Vielleicht kannst du einiges davon ins Projekt integrieren!
Schöne Grüße aus Innsbruck
Simon

riemarc committed:
Use readthedocs theme locally too

riemarc committed:
Added matplotlib animation/plotting window (only for animation purpose)

At the moment the new visualizer class MplVisualizer is
only used to distinguish between mpl and vtk visualizers.
But the class SimulationGui is furthermore to hard written
(for the vtk window).

riemarc committed:
Changed visualizer class structure

pymoskito.simulation_gui.SimulationGui is
still vtk-heavy

riemarc committed:
Added minimal matplotlib animation example

cklb committed:
Migration from PyQt4 to PyQt5

cklb committed:
Migrated missing dependencies and added Qt5 compatible interactor

cklb committed:
Added install instructions and migrated new example to coding standard of the package

cklb committed:
Merge branch 'riemarc-mpl_animation' into develop

cklb committed:
Changed signature of _control to better reflect the structure of the underlying control loop.

cklb committed:
Corrected minor error.

cklb committed:
Ran 2to3 conversion script.

cklb committed:
Fixed errors due to version conversion

Commit:
fix import errors

bu-ch committed:
Fix errors in pendulum example

cklb commented:
moin,

vielen Dank für deine Änderungsvorschläge! Wie sind gerade dabei diese einzuarbeiten und ich melde mich hier wenn es Neuigkeiten gibt.
damit es das nächste mal noch schneller geht, solltest du am besten das Repository in deinen Account forken, so kannst du nächstes mal ganz bequem einen PR stellen ohne den Code aufwändig kopieren zu müssen.

Gruß Stefan

cklb committed:
Added resources folder in setup.

cklb committed:
moved examples tests and resources under into directory.

cklb committed:
Added yaml as dependency and corrected MANIFEST.in to include config and simulation regime files

cklb committed:
Updated examples to be executable from installed module

cklb committed:
Extended export of simulation results with real file dialog

bu-ch committed:
Fix bug in ball and beam example, uncomment lss controller in ballbeam

bu-ch committed:
Integrate DeadTimeSensor improvements from pull request (niesi)

cklb committed:
added stuff

bu-ch committed:
Fix error in saving controller values

cklb committed:
Merge remote-tracking branch 'origin/bugfix-missing_debug_config' into bugfix-missing_debug_config

cklb committed:
Made vtk an optional dependency of all examples.

cklb committed:
Reformatted the Pendulum visualizer code.

cklb committed:
Corrected syntax error due to updated imports.

cklb committed:
Corrected imports

cklb committed:
Merge remote-tracking branch 'origin/develop' into develop

# Conflicts:
#	examples/balltube/main.py
#	examples/pendulum/visualization.py

cklb committed:
Changed behaviour of simulation_core to catch all exceptions during runtime, exit gracefully and forward them to the gui, so they can be shown to the user.

bu-ch committed:
dicts and lists are now accepted in settings

bu-ch committed:
Rework get_settings-function, accepted input types are standard python types, exponential function through exp( ) oder e^( ) and the number "pi" through pi, PI or Pi;
you can not use numpy! if you need an array, use nested lists and convert it in your own modules in an numpy-array

cklb committed:
Little text corrections.

bu-ch committed:
Fix QWheelEvent error (occurs if you zoom in a vtk window) by overloading the QVTKRenderWindowInteractor

bu-ch committed:
add matplot visualization for ball in tube example

bu-ch committed:
clean model and settings, add comments

bu-ch committed:
clean controller and feedforward, add regime examples

bu-ch committed:
linearise ball in tube system, check controllablity

bu-ch committed:
revised the two pendulum example, added matplotlib illustration

cklb commented:
Bis auf die Rückgabe von dictionaries von den einzelnen Simulationsmodulen wurden alle Änderungen soweit eingepflegt

cklb committed:
Updated requirements to new pyqtgraph release.

From now on, PyQt4 can be present in the runtime environment with pyqtgraph importing the wrong version of Qt. This make sthings a lot easier.

cklb committed:
Fixed bug that was created by "optimizing" the import in __init__.py ....

cklb committed:
Cleanup for generic PostProcessors.

bu-ch committed:
updated doc config and added documentation for pymoskito.tools

bu-ch committed:
revised tests and added some new tests for tool-functions, delete pytrajectory as dependency,

bu-ch committed:
Merge remote-tracking branch 'origin/develop' into f-test_cases

# Conflicts:
#	pymoskito/__init__.py

cklb committed:
Minor coding style changes

bu-ch committed:
fix bug:  lists with complex numbers in settings or sreg files were converted to a lists with strings

(cherry picked from commit bfbc662)
